modules/seqexec/server/src/main/scala/seqexec/server/gcal/GcalControllerEpics.scala

```diff
@@ -67,104 +41,93 @@ object GcalControllerEpics {
-  )
-
-  private def setThArLampParams(v: BinaryOnOff): List[IO[Unit]] = List(
-    GcalEpics.instance.lampsCmd.setThArLampName("ThAr"),
-    GcalEpics.instance.lampsCmd.setThArLampOn(v)
-  )
+  def apply[F[_]: Async](epics: => GcalEpics[F]): GcalController[F] = new GcalController[F] {
+    override def applyConfig(config: GcalConfig): F[Unit] =
+      retrieveConfig(epics).flatMap(configure(epics, _, config))
+  }
 
-  private def setQHLampParams(v: BinaryOnOff): List[IO[Unit]] = List(
-    GcalEpics.instance.lampsCmd.setQHLampName("QH"),
-    GcalEpics.instance.lampsCmd.setQHLampOn(v)
+  final case class EpicsGcalConfig(
+                                    lampAr: BinaryOnOff,
+                                    lampCuAr: BinaryOnOff,
+                                    lampQh: BinaryOnOff,
+                                    lampThAr: BinaryOnOff,
+                                    lampXe: BinaryOnOff,
+                                    lampIr: BinaryOnOff,
+                                    shutter: String,
+                                    filter: String,
+                                    diffuser: String
+                                  )
+
+  def retrieveConfig[F[_]: Async](epics: GcalEpics[F]): F[EpicsGcalConfig] = for {
+    ar   <- epics.lampAr
+    cuAr <- epics.lampCuAr
+    qh   <- epics.lampQH
+    thAr <- epics.lampThAr
+    xe   <- epics.lampXe
+    ir   <- epics.lampIr
+    shut <- epics.shutter
+    filt <- epics.filter
+    diff <- epics.diffuser
+  } yield EpicsGcalConfig(
+    ar,
+    cuAr,
+    qh,
+    thAr,
+    xe,
+    ir,
+    shut,
+    filt,
+    diff
   )
 
-  private def setXeLampParams(v: BinaryOnOff): List[IO[Unit]] = List(
-    GcalEpics.instance.lampsCmd.setXeLampName("Xe"),
-    GcalEpics.instance.lampsCmd.setXeLampOn(v)
-  )
+  def configure[F[_]: Async](epics: GcalEpics[F], current: EpicsGcalConfig, demand: GcalConfig): F[Unit] = {
+    val params: List[F[Unit]] = List(
+      applyParam(current.lampAr, encode(demand.lampAr.self), setArLampParams(epics)),
+      applyParam(current.lampCuAr, encode(demand.lampCuAr.self), setCuArLampParams(epics)),
+      applyParam(current.lampQh, encode(demand.lampQh.self), setQHLampParams(epics)),
+      applyParam(current.lampThAr, encode(demand.lampThAr.self), setThArLampParams(epics)),
+      applyParam(current.lampXe, encode(demand.lampXe.self), setXeLampParams(epics)),
+      demand.lampIrO.flatMap(d => applyParam(current.lampIr, encode(d.self), setIrLampParams(epics))),
+      applyParam(current.shutter, encode(demand.shutter), epics.shutterCmd.setPosition),
+      demand.filterO.flatMap(d => applyParam(current.filter, encode(d), epics.filterCmd.setName)),
+      demand.diffuserO.flatMap(d => applyParam(current.diffuser, encode(d), epics.diffuserCmd.setName))
+    ).flattenOption
+
+    (
+      params.sequence *>
+        epics.lampsCmd.setTimeout[F](SetupTimeout) *>
+        epics.post
+    ).whenA(params.nonEmpty)
```

Comment:
I always think we should go to make `applyParam` to return `F[Unit]` rather than `F[Option[Unit]]` but we need to skip the `setTimout/post` steps if no param is set. Is that correct?

Reply:
Yes, that is correct. Triggering a configuration command when no parameter has changed is a waste of time.